pkg/resolver/cas/resolver.go

```diff
@@ -0,0 +1,126 @@
+//    passed in. This data will then get stored in the local CAS. Finally, the data is returned to the caller.
+// In both cases above, the CID produced by the local CAS will be checked against the cid passed in to ensure they are
+// the same.
+func (h *Resolver) Resolve(webCASURL *url.URL, cid string, data []byte) ([]byte, error) {
+	if data != nil {
+		err := h.storeLocallyAndVerifyCID(data, cid)
+		if err != nil {
+			return nil, fmt.Errorf("failure while storing the data in the local CAS: %w", err)
+		}
+
+		return data, nil
+	}
+
+	// Ensure we have the data stored in the local CAS.
+	dataFromLocal, err := h.localCAS.Read(cid)
+	if err != nil {
+		if errors.Is(err, cas.ErrContentNotFound) {
+			dataFromRemote, errGetAndStoreRemoteData := h.getAndDataFromRemote(webCASURL, cid)
+			if errGetAndStoreRemoteData != nil {
+				return nil, fmt.Errorf("failure while getting and storing data from the remote "+
+					"WebCAS endpoint: %w", errGetAndStoreRemoteData)
+			}
+
+			return dataFromRemote, nil
+		}
+
+		return nil, fmt.Errorf("unexpected failure while checking local CAS for data stored at %s: %w",
+			cid, err)
+	}
+
+	return dataFromLocal, nil
+}
+
+func (h *Resolver) getAndDataFromRemote(webCASEndpoint *url.URL, cid string) ([]byte, error) {
+	resp, err := h.httpClient.Get(webCASEndpoint.String())
+	if err != nil {
+		return nil, fmt.Errorf("failed to execute GET call on %s: %w", webCASEndpoint.String(), err)
+	}
+
+	defer func() {
+		errClose := resp.Body.Close()
+		if errClose != nil {
+			logger.Errorf("failed to close response body from WebCAS endpoint: %s", errClose.Error())
+		}
+	}()
+
+	responseBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body from remote WebCAS endpoint: %w", err)
+	}
+
+	if resp.StatusCode == http.StatusOK {
+		errStoreLocallyAndVerifyCID := h.storeLocallyAndVerifyCID(responseBody, cid)
+		if errStoreLocallyAndVerifyCID != nil {
+			return nil, fmt.Errorf("failure while storing data retrieved from the remote "+
+				"WebCAS endpoint locally: %w", errStoreLocallyAndVerifyCID)
+		}
+	} else {
+		return nil, fmt.Errorf("failed to retrieve data from %s. "+
+			"Response status code: %d. Response body: %s", webCASEndpoint, resp.StatusCode, string(responseBody))
+	}
+
+	return responseBody, nil
+}
+
+func (h *Resolver) storeLocallyAndVerifyCID(data []byte, cidFromOriginalRequest string) error {
+	newCIDFromLocalCAS, err := h.localCAS.Write(data)
+	if err != nil {
+		return fmt.Errorf("failed to write data to CAS "+
+			"(and calculate CID in the process of doing so): %w", err)
+	}
+
+	logger.Debugf("Successfully stored data into CAS. Request CID [%s], "+
+		"CID as determined by local store [%s], Data: %s", cidFromOriginalRequest, newCIDFromLocalCAS,
+		string(data))
+
+	if newCIDFromLocalCAS != cidFromOriginalRequest {
+		return fmt.Errorf("successfully stored data into the local CAS, but the CID produced by "+
```

Comment:
It would be nice to verify the CID before storing, but at this point we don't know what CID version the CAS store is using.

Reply:
Unfortunately I don't really have this ability with the current interface setup and with how the IPFS client works. The IPFS client does the CID+storage in one call (the server returns the CID), and so the CAS client just has Write and Read as part of the interface. We'd need some sort of intermediate step

Reply:
Created #344 to enable support for both v0 and v1 CIDs (when possible)